File: tools/usage_window.py
#!/usr/bin/env python3
"""Turn verification demand into the next mining window.

Reads a <base>.usage.jsonl ledger (written by CnetMcpServer per
cnet_verify_claim call) and emits a CNET_WINDOW_FILE of the most-requested
token ids the base could NOT serve — uncovered claim tokens, weighted by
request frequency. Already-certified ids (any --exclude window/ledger) are
filtered out, so successive windows grow coverage instead of re-mining it.

Usage:
  usage_window.py <usage.jsonl> [-V 256] [--exclude window.txt ...]
                  [--tokenizer-dir DIR] [-o out.txt]

Output: one token id per line (mine with CNET_WINDOW_FILE=<out>), plus a
human-readable .words.txt sidecar when a tokenizer.json is available.
"""

import argparse
import json
import sys
from collections import Counter
from pathlib import Path

DEFAULT_TOKENIZER_DIR = "/home/marble/AI/Models/gemma-4-31B-it-int4-AutoRound"


def load_excludes(paths):
    seen = set()
    for p in paths:
        for line in Path(p).read_text().splitlines():
            line = line.strip()
            if not line:
                continue
            tok = line.split()[0]
            if tok.isdigit():
                seen.add(int(tok))
            elif tok.startswith("tk"):  # gaps ledger form tk<id>q<id>
                digits = tok[2:].split("q")[0]
                if digits.isdigit():
                    seen.add(int(digits))
    return seen


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("usage", help="<base>.usage.jsonl ledger")
    ap.add_argument("-V", type=int, default=256, help="window size")
    ap.add_argument("--exclude", action="append", default=[],
                    help="window file or gaps ledger of already-mined ids")
    ap.add_argument("--tokenizer-dir", default=DEFAULT_TOKENIZER_DIR)
    ap.add_argument("-o", "--out", default=None)
    args = ap.parse_args()

    demand = Counter()
    lines = 0
    for raw in Path(args.usage).read_text().splitlines():
        raw = raw.strip()
        if not raw:
            continue
        try:
            entry = json.loads(raw)
        except json.JSONDecodeError:
            continue
        lines += 1
        for tid in entry.get("uncovered", []):
            demand[int(tid)] += 1

    excluded = load_excludes(args.exclude)
    ranked = [(tid, n) for tid, n in demand.most_common() if tid not in excluded]
    if len(ranked) < args.V:
        print(f"note: only {len(ranked)} uncovered ids in demand "
              f"(from {lines} verifications); window will be short of "
              f"V={args.V} — mine it with V={len(ranked)}", file=sys.stderr)
    picked = ranked[: args.V]
    if not picked:
        print("no uncovered demand recorded — nothing to mine", file=sys.stderr)
        return 1

    out = Path(args.out or (str(Path(args.usage)).replace(
        ".usage.jsonl", "") + f".demand_window_{len(picked)}.txt"))
    out.write_text("".join(f"{tid}\n" for tid, _ in picked))

    # words sidecar when the tokenizer vocab is at hand
    tok_json = Path(args.tokenizer_dir) / "tokenizer.json"
    if tok_json.exists():
        vocab = json.loads(tok_json.read_text())["model"]["vocab"]
        id2tok = {v: k for k, v in vocab.items()}
        side = out.with_suffix(".words.txt")
        side.write_text("".join(
            f"{tid}\t{id2tok.get(tid, '?')}\t{n}\n" for tid, n in picked))
        print(f"wrote {out} and {side}")
    else:
        print(f"wrote {out}")
    print(f"demand: {len(demand)} distinct uncovered ids over {lines} "
          f"verifications; window: {len(picked)}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: dotnet/CnetControlPlane/Usage/UsageWindow.cs
using System.Text.Json;

namespace CnetControlPlane.Usage;

public static class UsageWindow
{
    public const string DefaultTokenizerDir = "/home/marble/AI/Models/gemma-4-31B-it-int4-AutoRound";

    public static HashSet<int> LoadExcludes(IEnumerable<string> paths)
    {
        var seen = new HashSet<int>();
        foreach (var p in paths)
        {
            foreach (var line in File.ReadAllLines(p))
            {
                var trimmed = line.Trim();
                if (trimmed.Length == 0)
                    continue;
                var tok = trimmed.Split(' ', StringSplitOptions.RemoveEmptyEntries)[0];
                if (int.TryParse(tok, out var id))
                    seen.Add(id);
                else if (tok.StartsWith("tk", StringComparison.Ordinal))
                {
                    var digits = tok[2..].Split('q')[0];
                    if (int.TryParse(digits, out var tid))
                        seen.Add(tid);
                }
            }
        }
        return seen;
    }

    public static (List<(int Tid, int Count)> Picked, int DemandDistinct, int Lines, string OutPath) BuildWindow(
        string usagePath,
        int windowSize,
        IEnumerable<string> excludes,
        string? outPath,
        string tokenizerDir)
    {
        var demand = new Dictionary<int, int>();
        var lines = 0;
        foreach (var raw in File.ReadAllLines(usagePath))
        {
            if (string.IsNullOrWhiteSpace(raw))
                continue;
            try
            {
                using var doc = JsonDocument.Parse(raw);
                lines++;
                if (!doc.RootElement.TryGetProperty("uncovered", out var uncovered)
                    || uncovered.ValueKind != JsonValueKind.Array)
                {
                    continue;
                }
                foreach (var tid in uncovered.EnumerateArray())
                {
                    var id = tid.GetInt32();
                    demand[id] = demand.GetValueOrDefault(id) + 1;
                }
            }
            catch (JsonException)
            {
                /* skip */
            }
        }

        var excluded = LoadExcludes(excludes);
        var ranked = demand.OrderByDescending(kv => kv.Value)
            .ThenBy(kv => kv.Key)
            .Where(kv => !excluded.Contains(kv.Key))
            .Select(kv => (kv.Key, kv.Value))
            .ToList();
        if (ranked.Count < windowSize)
        {
            Console.Error.WriteLine(
                $"note: only {ranked.Count} uncovered ids in demand "
                + $"(from {lines} verifications); window will be short of "
                + $"V={windowSize} — mine it with V={ranked.Count}");
        }
        var picked = ranked.Take(windowSize).ToList();
        if (picked.Count == 0)
            throw new InvalidOperationException("no uncovered demand recorded — nothing to mine");

        var outFile = outPath ?? usagePath.Replace(".usage.jsonl", "") + $".demand_window_{picked.Count}.txt";
        File.WriteAllText(outFile, string.Concat(picked.Select(p => $"{p.Key}\n")));

        var tokJson = Path.Combine(tokenizerDir, "tokenizer.json");
        if (File.Exists(tokJson))
        {
            using var doc = JsonDocument.Parse(File.ReadAllText(tokJson));
            var vocab = doc.RootElement.GetProperty("model").GetProperty("vocab");
            var id2tok = new Dictionary<int, string>();
            foreach (var prop in vocab.EnumerateObject())
                id2tok[prop.Value.GetInt32()] = prop.Name;
            var side = Path.ChangeExtension(outFile, ".words.txt");
            if (side == outFile)
                side = outFile + ".words.txt";
            // Python: out.with_suffix(".words.txt") — for .txt becomes .words.txt
            side = Path.Combine(
                Path.GetDirectoryName(outFile) ?? ".",
                Path.GetFileNameWithoutExtension(outFile) + ".words.txt");
            File.WriteAllText(side, string.Concat(
                picked.Select(p => $"{p.Key}\t{id2tok.GetValueOrDefault(p.Key, "?")}\t{p.Value}\n")));
            Console.WriteLine($"wrote {outFile} and {side}");
        }
        else
        {
            Console.WriteLine($"wrote {outFile}");
        }
        Console.WriteLine(
            $"demand: {demand.Count} distinct uncovered ids over {lines} "
            + $"verifications; window: {picked.Count}");
        return (picked, demand.Count, lines, outFile);
    }

    public static int RunCli(string[] args)
    {
        if (args.Length == 0)
        {
            Console.Error.WriteLine("usage: usage-window <usage.jsonl> [-V 256] [--exclude ...] [--tokenizer-dir DIR] [-o out]");
            return 2;
        }
        var usage = args[0];
        var window = 256;
        var excludes = new List<string>();
        var tokenizerDir = DefaultTokenizerDir;
        string? outPath = null;
        for (var i = 1; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "-V": window = int.Parse(args[++i]); break;
                case "--exclude": excludes.Add(args[++i]); break;
                case "--tokenizer-dir": tokenizerDir = args[++i]; break;
                case "-o":
                case "--out": outPath = args[++i]; break;
                default: throw new ArgumentException($"unknown argument: {args[i]}");
            }
        }
        try
        {
            BuildWindow(usage, window, excludes, outPath, tokenizerDir);
            return 0;
        }
        catch (InvalidOperationException ex)
        {
            Console.Error.WriteLine(ex.Message);
            return 1;
        }
    }
}
